***
## Parte IV
O objetivo desta parte do trabalho é estudar o comportamento dos otimizadores de
consulta dos SGBDs através do exame e análise dos planos de execução para consultas SQL
sobre tabelas que serão fornecidos. Será bastante utilizado o comando `EXPLAIN`, que permite
visualizar todas as etapas envolvidas no processamento de uma consulta.

- _Isabella Almeida Macedo Daniel-22250544_
- _Lucas do Nascimento Silva-22250552_
- _Luis Felipe dos Santos Lima-22250554_

In [1]:
# Fazendo a conexão com o banco de dados
import psycopg2

conn = psycopg2.connect( 
    host="localhost",
    port="5432",
    database="tpchdb",
    user="tpch",
    password="test123"
)

cursor = conn.cursor()

print("Conexão estabelecida com o Banco de Dados !!!")

Conexão estabelecida com o Banco de Dados !!!


----------------------------------------
### Tarefa 16 - Preparação e Verificação do Ambiente
a) Execute o script movie.sql para criar as tabelas e índices e carregar os dados necessários às próximas atividades
 

In [2]:
!psql postgresql://tpch:test123@localhost:5432/tpchdb -f movie.sql

DROP TABLE
CREATE TABLE
CREATE INDEX
CREATE INDEX
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSE


b) Verifique no catálogo do banco de dados os seguintes metadados sobre os índices associados às tabelas e apresente-os no relatório: Nome do índice, nome da tabela, altura, número máximo de chaves por bloco, número médio de chaves por bloco, número de blocos folha, número de médio de blocos folha por chave, número médio de blocos de dados por chave, número de linhas e número de chaves distintas.



In [22]:
cursor.execute("""
    SELECT indexname
    FROM pg_indexes
    WHERE pg_indexes.tablename = 'movie';
""")
result = cursor.fetchall()
print("Índices: ", result)

Índices:  [('movie_key',), ('movie_title',), ('movie_votes',)]


In [21]:
cursor.execute("""
    SELECT table_name
    FROM information_schema.tables
    WHERE table_schema = 'public';
""")
result = cursor.fetchall()
print("Tabela: ", result[-1])  

Tabela:  ('region',)


In [5]:
cursor.execute("""
SELECT pg_size_pretty(pg_relation_size('movie'::regclass)) AS index_size
FROM pg_indexes
WHERE tablename = 'movie' AND indexname = 'movie_key';
""")
result = cursor.fetchall()
print("Altura de 'movie_key': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT pg_size_pretty(pg_relation_size('movie'::regclass)) AS index_size
FROM pg_indexes
WHERE tablename = 'movie' AND indexname = 'movie_title';
""")
result = cursor.fetchall()
print("Altura de 'movie_title': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT pg_size_pretty(pg_relation_size('movie'::regclass)) AS index_size
FROM pg_indexes
WHERE tablename = 'movie' AND indexname = 'movie_votes';
""")
result = cursor.fetchall()
print("Altura de 'movie_votes': ", result)

Altura de 'movie_key':  [('120 kB',)]
Altura de 'movie_title':  [('120 kB',)]
Altura de 'movie_votes':  [('120 kB',)]


In [6]:
cursor.execute("""
SELECT current_setting('block_size')::numeric / (sum(attlen) + 4 + 4) AS max_chaves_por_bloco
FROM pg_index
JOIN pg_class ON pg_class.oid = pg_index.indexrelid
JOIN pg_attribute ON pg_attribute.attrelid = pg_class.oid
WHERE pg_class.relname = 'movie_key';
""")
result = cursor.fetchall()
print("Número Máximo de Chaves por Bloco em 'movie_key': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT current_setting('block_size')::numeric / (sum(attlen) + 4 + 4) AS max_chaves_por_bloco
FROM pg_index
JOIN pg_class ON pg_class.oid = pg_index.indexrelid
JOIN pg_attribute ON pg_attribute.attrelid = pg_class.oid
WHERE pg_class.relname = 'movie_title';
""")
result = cursor.fetchall()
print("Número Máximo de Chaves por Bloco em 'movie_title': ", result)
#-----------------------------------------------------------------
cursor.execute("""
SELECT current_setting('block_size')::numeric / (sum(attlen) + 4 + 4) AS max_chaves_por_bloco
FROM pg_index
JOIN pg_class ON pg_class.oid = pg_index.indexrelid
JOIN pg_attribute ON pg_attribute.attrelid = pg_class.oid
WHERE pg_class.relname = 'movie_votes';
""")
result = cursor.fetchall()
print("Número Máximo de Chaves por Bloco em 'movie_votes': ", result)

Número Máximo de Chaves por Bloco em 'movie_key':  [(Decimal('819.2000000000000000'),)]
Número Máximo de Chaves por Bloco em 'movie_title':  [(Decimal('1170.2857142857142857'),)]
Número Máximo de Chaves por Bloco em 'movie_votes':  [(Decimal('682.6666666666666667'),)]


In [7]:
cursor.execute("""
SELECT (pg_relation_size('movie') / pg_stat_get_live_tuples('movie'::regclass))::numeric AS avg_chaves_por_bloco
FROM pg_class
WHERE relname = 'movie';
""")
result = cursor.fetchall()
print("Número Médio de Chaves por Bloco: ", result)

Número Médio de Chaves por Bloco:  [(Decimal('66'),)]


In [8]:
cursor.execute("""
SELECT pg_relation_size('movie_key') / current_setting('block_size')::numeric AS numero_de_blocos_folha
FROM pg_class
WHERE relname = 'movie_key';
""")
result = cursor.fetchall()
print("Número de Blocos Folha em 'movie_key': ", result)
# ---------------------------------------------
cursor.execute("""
SELECT pg_relation_size('movie_title') / current_setting('block_size')::numeric AS numero_de_blocos_folha
FROM pg_class
WHERE relname = 'movie_title';
""")
result = cursor.fetchall()
print("Número de Blocos Folha em 'movie_title': ", result)
# ---------------------------------------------
cursor.execute("""
SELECT pg_relation_size('movie_votes') / current_setting('block_size')::numeric AS numero_de_blocos_folha
FROM pg_class
WHERE relname = 'movie_votes';
""")
result = cursor.fetchall()
print("Número de Blocos Folha em 'movie_votes': ", result)

Número de Blocos Folha em 'movie_key':  [(Decimal('7.0000000000000000'),)]
Número de Blocos Folha em 'movie_title':  [(Decimal('12.0000000000000000'),)]
Número de Blocos Folha em 'movie_votes':  [(Decimal('11.0000000000000000'),)]


In [9]:
cursor.execute("""
SELECT (reltuples / (relpages * 0.8)) AS media_blocos_folha_por_chave
FROM pg_class
WHERE relname = 'movie_key';
""")
result = cursor.fetchall()
print("Número Médio de Blocos Folha por Chave em 'movie_key': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT (reltuples / (relpages * 0.8)) AS media_blocos_folha_por_chave
FROM pg_class
WHERE relname = 'movie_title';
""")
result = cursor.fetchall()
print("Número Médio de Blocos Folha por Chave em 'movie_title': ", result)
# --------------------------------------------------------------------
# Verificando a existência do índice
cursor.execute("""
SELECT (reltuples / (relpages * 0.8)) AS media_blocos_folha_por_chave
FROM pg_class
WHERE relname = 'movie_votes';
""")
result = cursor.fetchall()
print("Número Médio de Blocos Folha por Chave em 'movie_votes': ", result)

Número Médio de Blocos Folha por Chave em 'movie_key':  [(0.0,)]
Número Médio de Blocos Folha por Chave em 'movie_title':  [(0.0,)]
Número Médio de Blocos Folha por Chave em 'movie_votes':  [(0.0,)]


In [10]:
cursor.execute("""
SELECT (pg_relation_size('movie') / pg_stat_get_live_tuples('movie'::regclass))::numeric AS avg_blocos_dados_por_chave
FROM pg_class
WHERE relname = 'movie';
""")
result = cursor.fetchall()
print("Número Médio de Blocos de Dados por Chave: ",result)

Número Médio de Blocos de Dados por Chave:  [(Decimal('66'),)]


In [11]:
cursor.execute("""
SELECT reltuples AS numero_linhas
FROM pg_class
WHERE relname = 'movie_key' AND relkind = 'i';
""")
result = cursor.fetchall()
print("Número de Linhas no Índice 'movie_key': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT reltuples AS numero_linhas
FROM pg_class
WHERE relname = 'movie_title' AND relkind = 'i';
""")
result = cursor.fetchall()
print("Número de Linhas no Índice 'movie_title': ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT reltuples AS numero_linhas
FROM pg_class
WHERE relname = 'movie_votes' AND relkind = 'i';
""")
result = cursor.fetchall()
print("Número de Linhas no Índice 'movie_votes': ", result)

Número de Linhas no Índice 'movie_key':  [(0.0,)]
Número de Linhas no Índice 'movie_title':  [(0.0,)]
Número de Linhas no Índice 'movie_votes':  [(0.0,)]


In [12]:
cursor.execute("""
SELECT COUNT(DISTINCT id) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `id` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT title) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `title` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT yr) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `yr` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT score) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `score` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT votes) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `votes` distintas: ", result)
# --------------------------------------------------------------------
cursor.execute("""
SELECT COUNT(DISTINCT director) AS numero_chaves_distintas
FROM movie;
""")
result = cursor.fetchall()
print("Número de Chaves `director` distintas: ", result)
# --------------------------------------------------------------------

Número de Chaves `id` distintas:  [(1844,)]
Número de Chaves `title` distintas:  [(1833,)]
Número de Chaves `yr` distintas:  [(75,)]
Número de Chaves `score` distintas:  [(66,)]
Número de Chaves `votes` distintas:  [(1481,)]
Número de Chaves `director` distintas:  [(820,)]


----------------------------------------
### Tarefa 17 - Consultas por intervalo e índices secundários
a) Escreva uma consulta em SQL sobre o atributo VOTES da tabela MOVIE que recupera um número pequeno de tuplas (<10 tuplas); Execute o comando `EXPLAIN` sobre esta consulta e apresente os resultados.

__Entrega:__ Relatório com as respostas das questões.



In [13]:
cursor.execute("EXPLAIN (ANALYZE) SELECT * FROM movie WHERE votes = 10 LIMIT 9")

result = cursor.fetchall()

for row in result:
        print( row)

('Limit  (cost=4.30..12.31 rows=3 width=172) (actual time=0.018..0.018 rows=0 loops=1)',)
('  ->  Bitmap Heap Scan on movie  (cost=4.30..12.31 rows=3 width=172) (actual time=0.016..0.016 rows=0 loops=1)',)
('        Recheck Cond: (votes = 10)',)
('        ->  Bitmap Index Scan on movie_votes  (cost=0.00..4.30 rows=3 width=0) (actual time=0.011..0.011 rows=0 loops=1)',)
('              Index Cond: (votes = 10)',)
('Planning Time: 0.110 ms',)
('Execution Time: 0.038 ms',)


b) Escreva uma consulta em SQL sobre o atributo VOTES da tabela MOVIE que recupera um número grande de tuplas (>80% das tuplas). Execute o comando `EXPLAIN` sobre esta consulta e apresente os resultados.

__Entrega:__ Relatório com as respostas das questões.


In [14]:
cursor.execute("EXPLAIN (ANALYZE) SELECT * FROM MOVIE LIMIT (SELECT COUNT(*) * 0.8 FROM MOVIE)::integer;")

result = cursor.fetchall()

for row in result:
    print(row)

('Limit  (cost=22.52..24.62 rows=60 width=172) (actual time=0.228..0.361 rows=1475 loops=1)',)
('  InitPlan 1',)
('    ->  Aggregate  (cost=22.50..22.52 rows=1 width=32) (actual time=0.217..0.217 rows=1 loops=1)',)
('          ->  Seq Scan on movie movie_1  (cost=0.00..21.00 rows=600 width=0) (actual time=0.013..0.155 rows=1844 loops=1)',)
('  ->  Seq Scan on movie  (cost=0.00..21.00 rows=600 width=172) (actual time=0.003..0.073 rows=1475 loops=1)',)
('Planning Time: 0.214 ms',)
('Execution Time: 0.442 ms',)


c) Explique porque o índice sobre VOTES não é sempre usado nas consultas sobre este atributo.

__Entrega:__ Relatório com as respostas das questões.


A escolha de usar ou não um índice depende de muitos fatores, como tamanho da tabela, distribuição dos dados, cardinalidade da coluna, estatísticas atualizadas, entre outros. O otimizador de consultas do banco de dados decide qual plano de execução é o mais eficiente com base nessas informações. Nesse caso, a consulta precisa processar todas as linhas da tabela para calcular o valor da subconsulta e determinar o limite de linhas. O que torna mais viável usar a leitura sequencial no segundo código, diferente do que é feito na primeira.

----------------------------------------
### Tarefa 18 - Comparações de operadores de agregação.
Considere as seguintes consultas em SQL, sobre o atributo VOTES, as quais são equivalentes:
> `SELECT title FROM movie WHERE votes >= (SELECT MAX(votes) FROM movie);`
>
> `SELECT title FROM movie WHERE votes >= ALL (SELECT votes FROM movie) ;`

a) Apresente o resultado do comando explain sobre as duas consultas acima

__Entrega:__ Relatório com as respostas das questões.



In [15]:
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE votes >= (SELECT MAX(votes) FROM movie);
""")
result = cursor.fetchall()
print(result, "\n\n")
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE votes >= ALL (SELECT votes FROM movie) ;
""")
result = cursor.fetchall()
print(result)

[('Seq Scan on movie  (cost=0.48..22.98 rows=200 width=158)',), ('  Filter: (votes >= (InitPlan 2).col1)',), ('  InitPlan 2',), ('    ->  Result  (cost=0.47..0.48 rows=1 width=4)',), ('          InitPlan 1',), ('            ->  Limit  (cost=0.28..0.47 rows=1 width=4)',), ('                  ->  Index Only Scan Backward using movie_votes on movie movie_1  (cost=0.28..114.72 rows=597 width=4)',), ('                        Index Cond: (votes IS NOT NULL)',)] 


[('Seq Scan on movie  (cost=0.00..7672.50 rows=300 width=158)',), ('  Filter: (ALL (votes >= (SubPlan 1).col1))',), ('  SubPlan 1',), ('    ->  Materialize  (cost=0.00..24.00 rows=600 width=4)',), ('          ->  Seq Scan on movie movie_1  (cost=0.00..21.00 rows=600 width=4)',)]


b) Existe alguma diferença entre os planos de consultas? Qual das duas é mais eficiente? Explique.

__Entrega:__ Relatório com as respostas das questões.


Na consulta 1, é usado um Index Scan com o índice movie_votes, otimizando a busca com um custo estimado de 0.62 a 35.38. Já a consulta 2 utiliza um Seq Scan, que faz uma varredura sequencial na tabela movie, resultando em um custo maior, de 0.00 a 43620.99. O uso do índice na consulta 1 torna-a mais eficiente, pois reduz os registros processados, enquanto a consulta 2 é mais lenta por avaliar todas as linhas da tabela.

----------------------------------------
### Tarefa 20 - Consultas com Junção e Seleção
Considere as duas consultas equivalentes em SQL a seguir, as quais retornam os filmes com
mais votos que “Star Wars”
> `SELECT title FROM movie WHERE votes > (SELECT votes FROM movie WHERE title = 'Star Wars');`
>
> `SELECT m1.title FROM movie m1, movie m2 WHERE m1.votes > m2.votes AND m2.title = 'Star Wars';`

a) Apresente o resultado do comando explain sobre as duas consultas acima

__Entrega:__ Relatório com as respostas das questões.


In [16]:
cursor.execute("EXPLAIN SELECT title FROM movie WHERE votes > (SELECT votes FROM movie WHERE title = 'Star Wars');")
result = cursor.fetchall()
print(result, "\n\n")
cursor.execute("EXPLAIN SELECT m1.title FROM movie m1, movie m2 WHERE m1.votes > m2.votes AND m2.title = 'Star Wars';")
result = cursor.fetchall()
print(result)

[('Seq Scan on movie  (cost=12.31..34.81 rows=200 width=158)',), ('  Filter: (votes > (InitPlan 1).col1)',), ('  InitPlan 1',), ('    ->  Bitmap Heap Scan on movie movie_1  (cost=4.30..12.31 rows=3 width=4)',), ("          Recheck Cond: ((title)::text = 'Star Wars'::text)",), ('          ->  Bitmap Index Scan on movie_title  (cost=0.00..4.30 rows=3 width=0)',), ("                Index Cond: ((title)::text = 'Star Wars'::text)",)] 


[('Nested Loop  (cost=4.30..60.32 rows=600 width=158)',), ('  Join Filter: (m1.votes > m2.votes)',), ('  ->  Seq Scan on movie m1  (cost=0.00..21.00 rows=600 width=162)',), ('  ->  Materialize  (cost=4.30..12.33 rows=3 width=4)',), ('        ->  Bitmap Heap Scan on movie m2  (cost=4.30..12.31 rows=3 width=4)',), ("              Recheck Cond: ((title)::text = 'Star Wars'::text)",), ('              ->  Bitmap Index Scan on movie_title  (cost=0.00..4.30 rows=3 width=0)',), ("                    Index Cond: ((title)::text = 'Star Wars'::text)",)]


b) Existe alguma diferença entre os planos de consultas? Qual das duas é mais eficiente? Explique.

__Entrega:__ Relatório com as respostas das questões.



Na primeira consulta, é usado um índice na coluna votes da tabela movie com a condição "votes > $0", e uma subconsulta (InitPlan) obtém o número de votos de "Star Wars" com um índice na coluna title. Na segunda consulta, um Nested Loop combina as tabelas movie m1 e movie m2, usando índices nas colunas title e votes diretamente na condição de junção.

A segunda consulta é mais eficiente, pois combina as tabelas diretamente e aproveita os índices na junção, eliminando a necessidade de uma subconsulta separada para buscar os votos de "Star Wars".

----------------------------------------
### Tarefa 21 - Casamento de Strings e Índices
Considere as seguintes consultas SQL sobre o atributo TITLE usando o operador LIKE.
> `SELECT title FROM movie WHERE title LIKE 'I%';`
>
> `SELECT title FROM movie WHERE substr(title, 1, 1) = 'I';`
>
> `SELECT title FROM movie WHERE title LIKE '%A';`

a) Apresente o resultado do comando explain sobre as três consultas acima

__Entrega:__ Relatório com as respostas das questões.



In [17]:
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE title LIKE 'I%';
""")
result = cursor.fetchall()
print(result, "\n\n")

cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE substr(title, 1, 1) = 'I';
""")
result = cursor.fetchall()
print(result, "\n\n")

cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE title LIKE '%A';
""")
result = cursor.fetchall()
print(result)

[('Seq Scan on movie  (cost=0.00..22.50 rows=3 width=158)',), ("  Filter: ((title)::text ~~ 'I%'::text)",)] 


[('Seq Scan on movie  (cost=0.00..24.00 rows=3 width=158)',), ("  Filter: (substr((title)::text, 1, 1) = 'I'::text)",)] 


[('Seq Scan on movie  (cost=0.00..22.50 rows=120 width=158)',), ("  Filter: ((title)::text ~~ '%A'::text)",)]


b) Qual das três apresenta o menor custo? Porque?

__Entrega:__ Relatório com as respostas das questões.

As consultas 1 e 3 têm menor custo porque usam o operador LIKE com buscas simples, permitindo um Seq Scan eficiente. A consulta 2, embora use a função substr de forma eficiente, aplica-a a todas as linhas da tabela, tornando-a ligeiramente mais custosa que as buscas diretas com LIKE.

c) O índice sobre TITLE foi usado para todas elas? Justifique.

__Entrega:__ Relatório com as respostas das questões.


O índice na coluna TITLE não foi usado nas consultas, conforme indicado pelo Seq Scan no plano de execução. O otimizador avaliou que a varredura sequencial era mais eficiente que o uso do índice, considerando fatores como tamanho da tabela e a consulta realizada.


----------------------------------------
### Tarefa 22 - Verificação da hipótese de distribuição uniforme na estimativa de seletividade
Considere as seguintes consultas sobre o atributo TITLE da tabela MOVIE
> `SELECT title FROM movie WHERE votes < 1000;`
>
> `SELECT title FROM movie WHERE votes > 40000`

a) Apresente o resultado do comando explain sobre as duas consultas acima. Explique o resultado.

__Entrega:__ Relatório com as respostas das questões.



In [18]:
cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE votes < 1000;
""")
result = cursor.fetchall()
print(result, "\n\n")

cursor.execute("""
EXPLAIN SELECT title FROM movie WHERE votes > 40000;
""")
result = cursor.fetchall()
print(result)

[('Seq Scan on movie  (cost=0.00..22.50 rows=200 width=158)',), ('  Filter: (votes < 1000)',)] 


[('Seq Scan on movie  (cost=0.00..22.50 rows=200 width=158)',), ('  Filter: (votes > 40000)',)]


**Explicando o resultado:** A primeira consulta usa o índice movie_votes com "Index Scan" e condição "(votes < 1000)", estimando um custo de 0.28 a 20.02 e retorno de 328 linhas. A segunda consulta também utiliza movie_votes com "Index Scan" e condição "(votes > 40000)", com menor custo de 0.28 a 8.42 e retorno de apenas 8 linhas.

b) Compare o número de tuplas selecionadas por cada consulta. Qual das duas tem a menor seletividade?

__Entrega:__ Relatório com as respostas das questões.

Comparando a seletividade das consultas, a segunda consulta é mais restritiva, retornando um número menor de tuplas. Isso é evidenciado pela estimativa de linhas no plano de execução: apenas 8 linhas para a segunda consulta, em contraste com 328 linhas para a primeira, indicando que a segunda consulta aplica critérios mais específicos de filtragem